// @colyseus/core/build/router/index.mjs
// packages/core/src/router/index.ts
import { createRouter as createBetterCallRouter, createEndpoint } from "@colyseus/better-call";
import { toNodeHandler } from "@colyseus/better-call/node";
import "../Transport.mjs";
import { controller } from "../matchmaker/controller.mjs";
import pkg from "../../package.json" with { type: "json" };
import {
  createEndpoint as createEndpoint2,
  createMiddleware,
  createInternalContext
} from "@colyseus/better-call";
function bindRouterToTransport(transport, router) {
  router.addEndpoint(createEndpoint("/__healthcheck", { method: "GET" }, async (ctx) => {
    return new Response("OK", { status: 200 });
  }));
  const expressApp = transport.getExpressApp();
  const hasRootRoute = (
    // check if express app has a root route
    expressRootRoute(expressApp) !== void 0 || // check if router has a root route
    Object.values(router.endpoints).some((endpoint) => endpoint.path === "/")
  );
  if (!hasRootRoute) {
    router.addEndpoint(createEndpoint("/", { method: "GET" }, async (ctx) => {
      return new Response(`Colyseus ${pkg.version}`, { status: 200 });
    }));
  }
  const server = transport.server;
  if (!server && transport.bindRouter) {
    transport.bindRouter(router);
    return;
  }
  let next = toNodeHandler(router.handler);
  if (expressApp) {
    server.removeListener("request", expressApp);
    expressApp.use(next);
    next = expressApp;
  }
  server.prependListener("request", (req, res) => {
    const corsHeaders = {
      ...controller.DEFAULT_CORS_HEADERS,
      ...controller.getCorsHeaders(new Headers(req.headers))
    };
    if (req.method === "OPTIONS") {
      res.writeHead(204, corsHeaders);
      res.end();
      return;
    }
    Object.entries(corsHeaders).forEach(([key, value]) => {
      res.setHeader(key, value);
    });
    next(req, res);
  });
}
function expressRootRoute(expressApp) {
  const stack = expressApp?._router?.stack ?? expressApp?.router?.stack;
  if (!stack) {
    return false;
  }
  return stack.find((layer) => layer.match("/") && !["query", "expressInit"].includes(layer.name));
}
var __globalEndpoints = {};
function createRouter(endpoints, config = {}) {
  __globalEndpoints = endpoints;
  return createBetterCallRouter({ ...endpoints }, config);
}
export {
  __globalEndpoints,
  bindRouterToTransport,
  createEndpoint2 as createEndpoint,
  createInternalContext,
  createMiddleware,
  createRouter,
  toNodeHandler
};
