// @colyseus/core/build/utils/Cloud.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/core/src/utils/Cloud.ts
var Cloud_exports = {};
__export(Cloud_exports, {
  checkInactiveSocketFile: () => checkInactiveSocketFile,
  getCloudConfig: () => getCloudConfig,
  getCloudPort: () => getCloudPort,
  getCloudSocketPath: () => getCloudSocketPath,
  isColyseusCloud: () => isColyseusCloud,
  shouldUseRedisConfig: () => shouldUseRedisConfig
});
module.exports = __toCommonJS(Cloud_exports);
var import_os = __toESM(require("os"), 1);
var import_fs = __toESM(require("fs"), 1);
var import_net = __toESM(require("net"), 1);
var import_Logger = require("../Logger.cjs");
var import_LocalDriver = require("../matchmaker/LocalDriver/LocalDriver.cjs");
var import_LocalPresence = require("../presence/LocalPresence.cjs");
var RedisDriver = import("@colyseus/redis-driver");
RedisDriver.catch(() => {
});
var RedisPresence = import("@colyseus/redis-presence");
RedisPresence.catch(() => {
});
function isColyseusCloud() {
  return process.env.COLYSEUS_CLOUD !== void 0;
}
function shouldUseRedisConfig() {
  return import_os.default.cpus().length > 1 || process.env.REDIS_URI !== void 0;
}
function getCloudPort(requestedPort) {
  let port = 2567;
  const processNumber = Number(process.env.NODE_APP_INSTANCE || "0");
  port += processNumber;
  return port;
}
function getCloudSocketPath(port) {
  return `/run/colyseus/${port}.sock`;
}
function checkInactiveSocketFile(sockFilePath) {
  return new Promise((resolve, reject) => {
    const client = import_net.default.createConnection({ path: sockFilePath }).on("connect", () => {
      client.end();
      reject(new Error(`EADDRINUSE: Already listening on '${sockFilePath}'`));
    }).on("error", () => {
      import_fs.default.unlink(sockFilePath, () => resolve(true));
    });
  });
}
async function getCloudConfig(port, currentDriver, currentPresence) {
  if (!isColyseusCloud()) {
    return null;
  }
  if (!shouldUseRedisConfig()) {
    return null;
  }
  let driver = currentDriver;
  let presence = currentPresence;
  const publicAddress = process.env.SUBDOMAIN + "." + process.env.SERVER_NAME + "/" + port;
  const needsRedisDriver = !driver || driver instanceof import_LocalDriver.LocalDriver;
  const needsRedisPresence = !presence || presence instanceof import_LocalPresence.LocalPresence;
  if (needsRedisDriver) {
    try {
      const module2 = await RedisDriver;
      driver = new module2.RedisDriver(process.env.REDIS_URI);
    } catch (e) {
      import_Logger.logger.error(e);
      import_Logger.logger.warn("");
      import_Logger.logger.warn("\u274C Could not initialize RedisDriver for Colyseus Cloud.");
      import_Logger.logger.warn("\u{1F449} npm install --save @colyseus/redis-driver");
      import_Logger.logger.warn("");
    }
  }
  if (needsRedisPresence) {
    try {
      const module2 = await RedisPresence;
      presence = new module2.RedisPresence(process.env.REDIS_URI);
    } catch (e) {
      import_Logger.logger.error(e);
      import_Logger.logger.warn("");
      import_Logger.logger.warn("\u274C Could not initialize RedisPresence for Colyseus Cloud.");
      import_Logger.logger.warn("\u{1F449} npm install --save @colyseus/redis-presence");
      import_Logger.logger.warn("");
    }
  }
  return { driver, presence, publicAddress };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  checkInactiveSocketFile,
  getCloudConfig,
  getCloudPort,
  getCloudSocketPath,
  isColyseusCloud,
  shouldUseRedisConfig
});
