// @colyseus/core/build/utils/Cloud.mjs
// packages/core/src/utils/Cloud.ts
import os from "os";
import fs from "fs";
import net from "net";
import { logger } from "../Logger.mjs";
import { LocalDriver } from "../matchmaker/LocalDriver/LocalDriver.mjs";
import { LocalPresence } from "../presence/LocalPresence.mjs";
var RedisDriver = import("@colyseus/redis-driver");
RedisDriver.catch(() => {
});
var RedisPresence = import("@colyseus/redis-presence");
RedisPresence.catch(() => {
});
function isColyseusCloud() {
  return process.env.COLYSEUS_CLOUD !== void 0;
}
function shouldUseRedisConfig() {
  return os.cpus().length > 1 || process.env.REDIS_URI !== void 0;
}
function getCloudPort(requestedPort) {
  let port = 2567;
  const processNumber = Number(process.env.NODE_APP_INSTANCE || "0");
  port += processNumber;
  return port;
}
function getCloudSocketPath(port) {
  return `/run/colyseus/${port}.sock`;
}
function checkInactiveSocketFile(sockFilePath) {
  return new Promise((resolve, reject) => {
    const client = net.createConnection({ path: sockFilePath }).on("connect", () => {
      client.end();
      reject(new Error(`EADDRINUSE: Already listening on '${sockFilePath}'`));
    }).on("error", () => {
      fs.unlink(sockFilePath, () => resolve(true));
    });
  });
}
async function getCloudConfig(port, currentDriver, currentPresence) {
  if (!isColyseusCloud()) {
    return null;
  }
  if (!shouldUseRedisConfig()) {
    return null;
  }
  let driver = currentDriver;
  let presence = currentPresence;
  const publicAddress = process.env.SUBDOMAIN + "." + process.env.SERVER_NAME + "/" + port;
  const needsRedisDriver = !driver || driver instanceof LocalDriver;
  const needsRedisPresence = !presence || presence instanceof LocalPresence;
  if (needsRedisDriver) {
    try {
      const module = await RedisDriver;
      driver = new module.RedisDriver(process.env.REDIS_URI);
    } catch (e) {
      logger.error(e);
      logger.warn("");
      logger.warn("\u274C Could not initialize RedisDriver for Colyseus Cloud.");
      logger.warn("\u{1F449} npm install --save @colyseus/redis-driver");
      logger.warn("");
    }
  }
  if (needsRedisPresence) {
    try {
      const module = await RedisPresence;
      presence = new module.RedisPresence(process.env.REDIS_URI);
    } catch (e) {
      logger.error(e);
      logger.warn("");
      logger.warn("\u274C Could not initialize RedisPresence for Colyseus Cloud.");
      logger.warn("\u{1F449} npm install --save @colyseus/redis-presence");
      logger.warn("");
    }
  }
  return { driver, presence, publicAddress };
}
export {
  checkInactiveSocketFile,
  getCloudConfig,
  getCloudPort,
  getCloudSocketPath,
  isColyseusCloud,
  shouldUseRedisConfig
};
